model_incoming_transactions.go

```diff
@@ -322,42 +328,51 @@ func processIncomingTransaction(ctx context.Context, logClient zLogger.GormLogge
 			}
 
 			// Broadcast was successful, so the transaction was accepted by the network, continue processing like before
-			if logClient != nil {
-				logClient.Info(ctx, fmt.Sprintf("broadcast of transaction was successful using %s", provider))
-			}
+			logClient.Info(ctx, fmt.Sprintf("processIncomingTransaction(): broadcast of transaction %s was successful using %s. Incoming tx will be processed again.", incomingTx.ID, provider))
+
 			// allow propagation
 			time.Sleep(3 * time.Second)
-			if txInfo, err = incomingTx.Client().Chainstate().QueryTransactionFastest(
-				ctx, incomingTx.ID, chainstate.RequiredInMempool, defaultQueryTxTimeout,
-			); err != nil {
-				incomingTx.Status = statusReady
-				incomingTx.StatusMessage = "tx was not found on-chain, attempting to broadcast using provider: " + provider
-				_ = incomingTx.Save(ctx)
-				return err
-			}
-		} else {
-			// Actual error occurred
-			bailAndSaveIncomingTransaction(ctx, incomingTx, err.Error())
-			return err
+			return nil // reprocess it when triggering the task again
 		}
+
+		// Actual error occurred
+		bailAndSaveIncomingTransaction(ctx, incomingTx, err.Error())
+		return err
 	}
 
-	if logClient != nil {
-		logClient.Info(ctx, fmt.Sprintf("found incoming transaction %s in %s", incomingTx.ID, txInfo.Provider))
+	// validate txInfo
+	if txInfo.BlockHash == "" || txInfo.MerkleProof == nil || txInfo.MerkleProof.TxOrID == "" || len(txInfo.MerkleProof.Nodes) == 0 {
+		logClient.Warn(ctx, fmt.Sprintf("processIncomingTransaction(): txInfo for %s is invalid, will try again later", incomingTx.ID))
+
+		if incomingTx.client.IsDebug() {
+			txInfoJSON, _ := json.Marshal(txInfo) //nolint:nolintlint,nilerr // error is not needed
+			incomingTx.DebugLog(string(txInfoJSON))
+		}
+		return nil
 	}
 
-	// Create the new transaction model
-	transaction := newTransactionFromIncomingTransaction(incomingTx)
+	logClient.Info(ctx, fmt.Sprintf("found incoming transaction %s in %s", incomingTx.ID, txInfo.Provider))
 
-	// Get the transaction by ID
-	if tx, _ := getTransactionByID(
-		ctx, transaction.rawXpubKey, transaction.TransactionBase.ID, transaction.client.DefaultModelOptions()...,
-	); tx != nil {
-		transaction = tx
+	// Check if we have transaction in DB already
+	transaction, _ := getTransactionByID(
+		ctx, incomingTx.rawXpubKey, incomingTx.ID, incomingTx.client.DefaultModelOptions()...,
+	)
+
+	if transaction == nil {
+		// Create the new transaction model
+		transaction = newTransactionFromIncomingTransaction(incomingTx)
```

Comment:
Can we use `incomingTx.toTransactionDto()`?

Reply:
No. These two functions have different levels of abstraction, at least they should. Ideally I would like to create a separate structures to use outside of bux, e.g. in bux-server. Of course it's not so simple now and neither bux or bux-server are ready to this. 

But I agree that my intentions do not follow directly from the code now... If you think that current version is confusing I'll remove `.toTransactionDto()` function